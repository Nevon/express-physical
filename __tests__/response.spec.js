"use strict";
const R = require("ramda");

const response = require("../src/response/response");
const serialize = require("../src/response/serialize");
const InvalidHealthcheckResponse = require("../src/response/invalid-healthcheck-response");
const { types, severities } = require("../src/constants");

const baseResponse = {
  name: "Name",
  actionable: true,
  type: types.SELF,
  info: {
    foo: {
      barBaz: true
    }
  },
  link: "http://google.com"
};

const healthyResponse = Object.assign({}, baseResponse, {
  healthy: true
});

const unhealthyResponse = Object.assign({}, baseResponse, {
  healthy: false,
  severity: severities.CRITICAL,
  message: "Message"
});

describe("Response", () => {
  describe("Validation", () => {
    const requiredFields = ["name", "actionable", "type"];

    requiredFields.forEach(field => {
      test(`It should throw an error when ${field} is missing`, () => {
        expect(() => response(R.omit([field], healthyResponse))).toThrow(
          InvalidHealthcheckResponse
        );
      });
    });

    const infoFields = ["info", "additional_info"];
    infoFields.forEach(field => {
      test(`It should throw an error when ${field} is not an object`, () => {
        expect(() =>
          response(
            Object.assign({}, healthyResponse, { info: "not an object" })
          )
        ).toThrow(InvalidHealthcheckResponse);
      });

      test(`It should not throw an error when ${field} is omitted`, () => {
        expect(() => response(R.omit([field], healthyResponse))).not.toThrow();
      });

      test(`It should not throw an error when ${field} is an object`, () => {
        expect(() =>
          response(
            Object.assign({}, healthyResponse, { info: { foo: { bar: true } } })
          )
        ).not.toThrow();
      });
    });

    describe("When the check is healthy", () => {
      test("It should throw an error if severity is provided", () => {
        expect(() =>
          response(
            Object.assign({}, healthyResponse, { severity: severities.WARNING })
          )
        ).toThrow(InvalidHealthcheckResponse);
      });
    });

    describe("When the result is unhealthy", () => {
      const requiredFieldsWhenUnhealthy = ["severity", "message"];

      requiredFieldsWhenUnhealthy.forEach(field => {
        test(`It should throw an error when ${field} is missing`, () => {
          expect(() => response(R.omit([field], unhealthyResponse))).toThrow(
            InvalidHealthcheckResponse
          );
        });
      });
    });

    const dependentOnRequiredForTypes = [
      types.INFRASTRUCTURE,
      types.INTERNAL_DEPENDENCY,
      types.EXTERNAL_DEPENDENCY
    ];
    describe("dependentOn is required for type", () => {
      dependentOnRequiredForTypes.forEach(type => {
        test(type, () => {
          expect(() =>
            response(Object.assign({}, unhealthyResponse, { type }))
          ).toThrow(InvalidHealthcheckResponse);
        });
      });
    });

    describe("dependentOn should not be present for type", () => {
      R.keys(R.omit(dependentOnRequiredForTypes, types)).forEach(type => {
        test(type, () => {
          expect(() =>
            response(
              Object.assign({}, healthyResponse, {
                type,
                dependentOn: "foobar"
              })
            )
          ).toThrow(InvalidHealthcheckResponse);
        });
      });
    });
  });

  describe("serialization", () => {
    const serializeResponse = R.compose(serialize, response);

    test("should format dependentOn as an object", () => {
      const actual = serializeResponse(
        Object.assign({}, unhealthyResponse, {
          type: types.INFRASTRUCTURE,
          dependentOn: "foo"
        })
      );
      const expected = {};
      expected["service_name"] = "foo";

      expect(actual["dependent_on"]).toEqual(expected);
    });

    test("should copy info to additionalInfo", () => {
      const actual = serializeResponse(healthyResponse);

      expect(actual["additional_info"]).toEqual(actual.info);
    });

    test("should remove empty fields", () => {
      const fields = R.keys(
        serializeResponse(
          Object.assign({}, unhealthyResponse, { info: undefined, link: "" })
        )
      );

      expect(fields.includes("info")).toBe(false);
      expect(fields.includes("link")).toBe(false);
    });

    test("should not reformat the keys of the `info` object", () => {
      const info = {
        servicingAPIResponse: {
          status: 500
        }
      };
      const actual = serializeResponse(
        Object.assign({}, unhealthyResponse, { info })
      );

      expect(actual.info).toEqual(info);
    });
  });
});
