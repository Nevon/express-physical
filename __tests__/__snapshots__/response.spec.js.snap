// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response Validation It should throw an error when actionable is missing 1`] = `"Actionable is required when healthcheck is healthy at "actionable""`;

exports[`Response Validation It should throw an error when additional_info is not an object 1`] = `"Info should be an object at "info""`;

exports[`Response Validation It should throw an error when info is not an object 1`] = `"Info should be an object at "info""`;

exports[`Response Validation It should throw an error when name is missing 1`] = `"Name is required when healthcheck is healthy at "name""`;

exports[`Response Validation It should throw an error when type is missing 1`] = `"Type is required when healthcheck is healthy at "type""`;

exports[`Response Validation When the result is unhealthy It should throw an error when message is missing 1`] = `"Message is required when healthcheck is unhealthy at "message""`;

exports[`Response Validation When the result is unhealthy It should throw an error when severity is missing 1`] = `"Severity is required when healthcheck is unhealthy at "severity""`;

exports[`Response Validation dependentOn is required for type EXTERNAL_DEPENDENCY 1`] = `"DependentOn is required when type is one of INFRASTRUCTURE, INTERNAL_DEPENDENCY, EXTERNAL_DEPENDENCY at "dependentOn""`;

exports[`Response Validation dependentOn is required for type INFRASTRUCTURE 1`] = `"DependentOn is required when type is one of INFRASTRUCTURE, INTERNAL_DEPENDENCY, EXTERNAL_DEPENDENCY at "dependentOn""`;

exports[`Response Validation dependentOn is required for type INTERNAL_DEPENDENCY 1`] = `"DependentOn is required when type is one of INFRASTRUCTURE, INTERNAL_DEPENDENCY, EXTERNAL_DEPENDENCY at "dependentOn""`;

exports[`Response Validation dependentOn should not be present for type INTERNET_CONNECTIVITY 1`] = `"DependentOn should be omitted when type is one of SELF, METRICS, INTERNET_CONNECTIVITY at "dependentOn""`;

exports[`Response Validation dependentOn should not be present for type METRICS 1`] = `"DependentOn should be omitted when type is one of SELF, METRICS, INTERNET_CONNECTIVITY at "dependentOn""`;

exports[`Response Validation dependentOn should not be present for type SELF 1`] = `"DependentOn should be omitted when type is one of SELF, METRICS, INTERNET_CONNECTIVITY at "dependentOn""`;
